### PR TITLE
Architecture Chapter Update

### PR DESCRIPTION
**Pull request by arthur-here**
In this update we shift focus from the MVVM/Rx approach to the global problem of our applications: separation of the Domain and Presentation models. Please, check it out and give your feedback. 

Here is a link for a comfortable view of the chapter (with images and without diffs): https://github.com/uptechteam/ios-cookbook/blob/6cd53ee4f6754e7d0f970f6565d3211183952610/4-architecture.md

### CODE REVIEW

File: 4-architecture.md [hunk 1]
```diff
@@ -1,286 +1,234 @@
...
-      route: messageDetailsRoute
-    )
-  }
+At the time of writing this chapter, Swift doesn't support asynchronous programming primitives such as *async*/*await*. Developers have to use callbacks for any kind of asynchronous operation. This [doesn't always end up well](http://callbackhell.com).
+
+Instead, we can resort to 3rd party libraries, such as Rx. Among a lot of concepts, RxSwift brings *Single*, which also knows as a *Future* or a *Promise*.
+
+**Future** object represents the eventual completion (or failure) of an asynchronous operation, and its resulting value. Future is a proxy for a value not necessarily known when the promise is created.
+
+Features allows to replace these code:
```
Comment (ArtemMysik) [inline 4-architecture.md]: Probably I understood something wrong, but should it be `Futures` instead of `Features` here?

File: 4-architecture.md [hunk 2]
```diff
@@ -1,286 +1,234 @@
...
+
+Apart from that Rx provides a lot of useful operators, such as `flatMapLatest`, `debounce` or `distinctUntilChanged`. Be aware of them, but do not overuse. Most of the asynchronous code is located in the Domain Model, so only part of the Presentation Layer that should touch Rx is subscribing to the model updates notifications.
+
+## 7. Redux
+
+Redux is another useful approach in our toolbelt. Originally, Redux is an implementation of the unidirectional data flow architecture, where the whole app state is represented in a single big tree-object, and mutations to this object are limited by a distinct number of actions. 
+
+This idea is also applicable to the unit of any size: whole app, single module or even just a class. Redux is an implementation detail of a single module so it doesn't affect the system. We use Redux for objects with a lot of interdependent states, it allows us to ssynchronize state management in a single function `reducer`. For more information refer to [the Redux chapter](4-1-redux.md).
```
Comment (ArtemMysik) [inline 4-architecture.md]: `ssynchronize` typo

File: 4-architecture.md [hunk 3]
```diff
@@ -1,286 +1,234 @@
...
 
-```swift
-// TeamChatViewController.swift
+## 2. Rich Domain Model
 
-final class TeamChatViewController: UIViewController {
-    private lazy var teamChatView = TeamChatView.instantiateFromNib()
-    private lazy var viewModel = TeamChatViewModel()
-    private let disposeBag = DisposeBag()
+Make your Domain Model smart. Use [design patterns](https://en.wikipedia.org/wiki/Software_design_pattern#Classification_and_list), unleash your engineering creativity here but be consistent. Keep in mind [our definition of a "Good Architecture"](#what-is-a-%22good-architecture%22).
```
Comment (romanfurman6) [inline 4-architecture.md]: `our definition of a "Good Architecture"` this anchor doesn't work for me let's remove `"` (%22)

File: 4-architecture.md [hunk 4]
```diff
@@ -1,286 +1,234 @@
-# Architecture
+# Architectural Principles <!-- omit in toc -->
 
-By architecture here we mean a set of design patterns used across the entire application. We believe that code quality is much more important than specific architecture patterns.
+This chapter defines ground ideas and architectural principles we follow at Uptech.
 
-However we have a number of reasons to use same architecture for every project we build.
+- [Background](#background)
+  - [What is a &quot;Good Architecture&quot;?](#what-is-a-quotgood-architecturequot)
```
Comment (romanfurman6) [inline 4-architecture.md]: `what-is-a-quotgood-architecturequot`
let's remove `qout`, looks like broken 😿

File: 4-architecture.md [hunk 5]
```diff
@@ -1,286 +1,234 @@
...
+These requirements are the minimum for any application design.
 
-        let outputs = viewModel.makeOutputs(from: inputs)
+**Advice:**
 
-        outputs.props
-            .observeOn(MainScheduler.instance)
-            .bind { [weak self] in self?.render(props: $0) }
-            .disposed(by: disposeBag)
+Use **facades** to build a nice API for your Presentation layer to use. Facades helps hide complex object graph of Domain model behind a single object. Example:
```
Comment (romanfurman6) [inline 4-architecture.md]: I would get rid of "to use" at the end. :)
Also, let's add a URL `https://en.wikipedia.org/wiki/Facade_pattern` for `facades` word to be more clear.
Or if we don't want to have too much Wikipedia in our cookbook :D, we can rephrase it: "Use facade pattern ... for your Presentation layer."

File: 4-architecture.md [hunk 6]
```diff
@@ -1,286 +1,234 @@
...
-// ProfileDetailsViewController.swift
-...
+*Notice that all the methods which update the model don't return any value. Presentation layer doesn't care about a result of the model update and updates itself only in response to notifications.*
 
-init(profileId: String) {
-    viewModel = ProfileDetailsViewModel(profileId: profileId)
-    super.init(nibName: nil, bundle: nil)
-}
-```
+## 3. Don't fight the iOS SDK
```
Comment (romanfurman6) [inline 4-architecture.md]: 😻

**Comment by EvgenyMatviyenko** [comment]
Very nice!
I like the idea of facades for domain model. Lately I was building my view models following identical principles, they are sort of facades for business logic that also contain their own logic, primaraly state managment
One thing that I do not agree is use of Redux without SwiftUI, I do not see a case where we need to update entire screen UI every single action (what doesn't mean that there are no such cases, they should be very specific as for me). Updating only parts of UI after state updates is a way to go for 99% of time

**Comment by arthur-here** [comment]
Thanks everyone for the review and comments! ❤️ 

Next step is to update Redux chapter

**Pull request merged**